File: UE04_Labyrinth/molver.py
__author__ = "Karun Sandhu"

import time
from collections import deque


def dfs(maze: list[str], start: tuple[int, int]) -> list[tuple[int, int]] | None:
    stack: deque[tuple[tuple[int, int], list[tuple[int, int]]]] = deque(
        [(start, [start])]
    )
    visited: set[tuple[int, int]] = set()

    while stack:
        (x, y), path = stack.pop()

        if (x, y) in visited:
            continue
        visited.add((x, y))

        if maze[y][x] == "A":
            return path

        for nx, ny in [
            (x + 1, y),
            (x - 1, y),
            (x, y + 1),
            (x, y - 1),
        ]:
            if 0 <= ny < len(maze) and 0 <= nx < len(maze[0]):
                if maze[ny][nx] != "#":
                    stack.append(((nx, ny), path + [(nx, ny)]))

    return None


def bfs(maze: list[str], start: tuple[int, int]) -> list[tuple[int, int]] | None:
    queue: deque[tuple[tuple[int, int], list[tuple[int, int]]]] = deque(
        [(start, [start])]
    )
    visited: set[tuple[int, int]] = set()

    while queue:
        (x, y), path = queue.popleft()

        if (x, y) in visited:
            continue
        visited.add((x, y))

        for nx, ny in [
            (x + 1, y),
            (x - 1, y),
            (x, y + 1),
            (x, y - 1),
        ]:
            if 0 <= ny < len(maze) and 0 <= nx < len(maze[0]):
                if maze[ny][nx] != "#":
                    if maze[ny][nx] == "A":
                        return path + [(nx, ny)]
                    queue.append(((nx, ny), path + [(nx, ny)]))

    return None


def load_maze(path):
    with open(path) as f:
        return [line.strip() for line in f]


def mark_map(maze: list[str], path: list[tuple[int, int]]) -> list[str]:
    maze_copy = [list(row) for row in maze]
    for x, y in path:
        if maze_copy[y][x] not in ("S", "A"):
            maze_copy[y][x] = "."
    return ["".join(row) for row in maze_copy]


if __name__ == "__main__":
    maze_files = [
        ("L1", "./UE04_Labyrinth/l1.txt"),
        ("L2", "./UE04_Labyrinth/l2.txt"),
        ("L3", "./UE04_Labyrinth/l3.txt"),
    ]

    for label, path in maze_files:
        maze = load_maze(path)

        start = time.perf_counter()
        dfs_result = dfs(maze, (1, 1))  # or whatever your start coordinate is
        elapsed = time.perf_counter() - start

        if label == "L2" and dfs_result is not None:
            print("\n".join(mark_map(maze, dfs_result)))

        print(
            f"DFS {label} found path of length {len(dfs_result) if dfs_result else 'None'} "
            f"in {elapsed:.6f} seconds"
        )

        start = time.perf_counter()
        bfs_result = bfs(maze, (1, 1))  # or whatever your start coordinate is
        elapsed = time.perf_counter() - start

        if label == "L2" and bfs_result is not None:
            print("\n".join(mark_map(maze, bfs_result)))

        print(
            f"BFS {label} found path of length {len(bfs_result) if bfs_result else 'None'} "
            f"in {elapsed:.6f} seconds"
        )

File: UE04_Labyrinth/molver.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"time"
)

type Position struct {
	X int
	Y int
}

type Maze struct {
	start Position
	grid  []string
}

func bfs(maze Maze) []Position {
	visited := make(map[Position]bool)
	parent := make(map[Position]Position)

	queue := make([]Position, 0, 1024)
	queue = append(queue, maze.start)
	visited[maze.start] = true

	directions := []Position{
		{1, 0},
		{-1, 0},
		{0, 1},
		{0, -1},
	}

	var exit Position
	found := false
	head := 0

	for head < len(queue) {
		curr := queue[head]
		head++

		if maze.grid[curr.Y][curr.X] == 'A' {
			exit = curr
			found = true
			break
		}

		for _, d := range directions {
			nx, ny := curr.X+d.X, curr.Y+d.Y
			if ny < 0 || ny >= len(maze.grid) || nx < 0 || nx >= len(maze.grid[0]) {
				continue
			}
			if maze.grid[ny][nx] == '#' {
				continue
			}

			neighbor := Position{nx, ny}
			if visited[neighbor] {
				continue
			}

			visited[neighbor] = true
			parent[neighbor] = curr
			queue = append(queue, neighbor)
		}
	}

	if !found {
		return nil
	}

	path := []Position{}
	for curr := exit; curr != maze.start; curr = parent[curr] {
		path = append(path, curr)
	}
	path = append(path, maze.start)

	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}

	return path
}

func loadMaze(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func markMap(maze []string, path []Position) []string {
	mazeCopy := make([][]rune, len(maze))
	for i := range maze {
		mazeCopy[i] = []rune(maze[i])
	}

	for _, pos := range path {
		if mazeCopy[pos.Y][pos.X] != 'S' && mazeCopy[pos.Y][pos.X] != 'A' {
			mazeCopy[pos.Y][pos.X] = '.'
		}
	}

	result := make([]string, len(mazeCopy))
	for i, row := range mazeCopy {
		result[i] = string(row)
	}
	return result
}

func main() {
	mazeFiles := []string{"./UE04_Labyrinth/l1.txt", "./UE04_Labyrinth/l2.txt", "./UE04_Labyrinth/l3.txt"}

	for i, path := range mazeFiles {
		grid, err := loadMaze(path)
		if err != nil {
			panic(err)
		}

		start := time.Now()
		result := bfs(Maze{start: Position{X: 1, Y: 1}, grid: grid})
		elapsed := time.Since(start)

		fmt.Printf("BFS L%d found path of length %d in %s\n", i+1, len(result), elapsed)

		if i == 1 && result != nil {
			for _, line := range markMap(grid, result) {
				fmt.Println(line)
			}
		}
	}
}
